### DISCUSSION TITLE
[API Proposal]: Add keywords return to method parameters

### DISCUSSION THREAD

**Discussion by FalconWu2017** [dotnet/csharplang#6848] [Language Ideas]
### Background and motivation

Simplified method compilation。

### API Proposal

```csharp

    public int Do(int p) {
        //Do something like this
        p += 1;

        return p;
    }
    
// return describes this method will return p like Do.
    public int Do1(return int p) {
        //Do something like this
        p += 1;

        //return p;
    }
```

### API Usage

Nothing special！
```csharp

Console.WriteLine(Demo.Do(2));//3

Console.WriteLine(Demo.Do1(1));//2
```

### Risks

No risks  I think.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by FalconWu2017** [comment]
I don't have any write-permissions to add label to this issue.

**Comment by HaloFour** [comment]
I don't see this being very useful.  It's not that common to manipulate and return the parameters of a method directly.  It's often considered a bad practice to mutate method parameters.

**Comment by FalconWu2017** [comment]
@HaloFour 
I think it is useful when coding extension methods.Almost extension methods will return the parameter itself.
```c#
 /// <summary>
 /// 增加FalconCalom相关服务
 /// </summary>
 public static IServiceCollection AddFalconClaim(this IServiceCollection services, Action<FalconClaimOption>? optionBuilder = null) {
     //new option
     var option = new FalconClaimOption();
    // invoke
     optionBuilder?.Invoke(option);
     //  Do something else.......................

     //return 
     return services;
  }
```
1st:
This kind of method is almost the standard mode for me to write extension methods,It manipulate and return the parameters of a method directly. 
2nd:
At this time, I have two choices. First, I break the coding thought and write the return statement first and then come back to continue coding. Second,Ignore the red prompt until I finish coding and write the return statement. Any way,Continuous coding ideas are very important but red hints are very annoying. 
If can add the return flag before the parameter to tell the compiler that the parameter to be returned and eliminate the red prompt, it is very good.

**Reply by HaloFour** [reply]
> Almost extension methods will return the parameter itself.

That is not my experience with extension methods.  Take LINQ, most of the methods return the same type, but they return a different instance.

> Second,Ignore the red prompt until I finish coding and write the return statement.

That red prompt is a good thing as it indicates that you lack a control statement to leave the method.  If the return is implicitly added at the end of the method that would make it easy to miss a case in a branching logic where you have forgotten to return the appropriate value.

Having to flag the parameter doesn't seem different than writing the `return` statement first and then completing the remainder of the method body.  Either way you're going to have to stop and consider the relationship of the parameters to the return value.

**Reply by FalconWu2017** [reply]
> That is not my experience with extension methods. Take LINQ, most of the methods return the same type, but they return a different instance.     

Extending IServiceCollection is different. Generally, we do not need a new instance.    

>  If the return is implicitly added at the end of the method that would make it easy to miss a case in a branching logic where you have forgotten to return the appropriate value.   

We can do like this.
![image](https://user-images.githubusercontent.com/34147999/210317615-4df5b3da-e9e4-4c92-812f-8f7e55573132.png)  

> Either way you're going to have to stop and consider the relationship of the parameters to the return value.   

No,We don't need to consider so much, because the same parameter is always returned in this scheme.
